File: game/src/particles.py
"""Particle system for visual effects"""
import pygame
import math
import random
from typing import List, Tuple
from dataclasses import dataclass

@dataclass
class Particle:
    """A single particle"""
    x: float
    y: float
    vx: float
    vy: float
    lifetime: float
    max_lifetime: float
    size: float
    color: Tuple[int, int, int]
    fade: bool = True
    shrink: bool = False
    gravity: float = 0.0

class ParticleSystem:
    """Manages all particles in the game"""
    
    # Particle physics constants
    PARTICLE_DRAG = 0.98
    
    def __init__(self):
        self.particles: List[Particle] = []
    
    def update(self, dt: float):
        """Update all particles"""
        # Iterate backwards to safely remove particles
        for i in range(len(self.particles) - 1, -1, -1):
            particle = self.particles[i]
            particle.lifetime -= dt
            
            if particle.lifetime <= 0:
                self.particles.pop(i)
                continue
            
            # Update position
            particle.x += particle.vx * dt
            particle.y += particle.vy * dt
            
            # Apply gravity
            if particle.gravity != 0:
                particle.vy += particle.gravity * dt
            
            # Apply drag
            particle.vx *= self.PARTICLE_DRAG
            particle.vy *= self.PARTICLE_DRAG
    
    def render(self, surface: pygame.Surface, camera_x: float, camera_y: float):
        """Render all particles"""
        for particle in self.particles:
            screen_x = int(particle.x - camera_x)
            screen_y = int(particle.y - camera_y)
            
            # Skip if off screen
            if screen_x < -10 or screen_x > surface.get_width() + 10:
                continue
            if screen_y < -10 or screen_y > surface.get_height() + 10:
                continue
            
            # Calculate alpha based on lifetime
            alpha = 255
            if particle.fade:
                alpha = int(255 * (particle.lifetime / particle.max_lifetime))
                alpha = max(0, min(255, alpha))
            
            # Calculate size
            size = particle.size
            if particle.shrink:
                size = particle.size * (particle.lifetime / particle.max_lifetime)
                size = max(0.5, size)
            
            # Render particle
            color = particle.color
            if alpha < 250 and particle.fade:
                # Use alpha blending for fading particles
                temp_surface = pygame.Surface((int(size * 2 + 2), int(size * 2 + 2)), pygame.SRCALPHA)
                color_with_alpha = (*color, alpha)
                pygame.draw.circle(temp_surface, color_with_alpha, 
                                 (int(size + 1), int(size + 1)), int(size))
                surface.blit(temp_surface, (screen_x - int(size + 1), screen_y - int(size + 1)))
            else:
                # Direct rendering for fully opaque particles (faster)
                pygame.draw.circle(surface, color, (screen_x, screen_y), int(size))
    
    def create_weapon_fire_effect(self, x: float, y: float, angle: float, weapon_type: str):
        """Create particles for weapon firing"""
        if weapon_type == "laser":
            # Laser muzzle flash - bright and quick
            for _ in range(8):
                speed = random.uniform(30, 60)
                spread = random.uniform(-0.3, 0.3)
                particle_angle = angle + spread
                vx = math.cos(particle_angle) * speed
                vy = math.sin(particle_angle) * speed
                
                color = random.choice([
                    (255, 100, 100),
                    (255, 150, 150),
                    (255, 200, 200),
                ])
                
                particle = Particle(
                    x=x,
                    y=y,
                    vx=vx,
                    vy=vy,
                    lifetime=0.2,
                    max_lifetime=0.2,
                    size=random.uniform(2, 4),
                    color=color,
                    fade=True,
                    shrink=True
                )
                self.particles.append(particle)
        
        else:  # cannon
            # Cannon smoke and fire
            for _ in range(15):
                speed = random.uniform(20, 50)
                spread = random.uniform(-0.5, 0.5)
                particle_angle = angle + spread
                vx = math.cos(particle_angle) * speed
                vy = math.sin(particle_angle) * speed
                
                # Mix of fire and smoke colors
                if random.random() < 0.6:
                    color = random.choice([
                        (255, 200, 0),   # Yellow fire
                        (255, 150, 0),   # Orange fire
                        (255, 100, 0),   # Red-orange
                    ])
                else:
                    color = random.choice([
                        (100, 100, 100), # Gray smoke
                        (150, 150, 150), # Light smoke
                    ])
                
                particle = Particle(
                    x=x,
                    y=y,
                    vx=vx,
                    vy=vy,
                    lifetime=random.uniform(0.3, 0.5),
                    max_lifetime=0.5,
                    size=random.uniform(3, 6),
                    color=color,
                    fade=True,
                    shrink=False
                )
                self.particles.append(particle)
    
    def create_explosion(self, x: float, y: float, size: str = "medium"):
        """Create an explosion effect"""
        # Determine explosion parameters based on size
        if size == "small":
            num_particles = 20
            speed_range = (50, 150)
            size_range = (2, 5)
            lifetime_range = (0.3, 0.6)
        elif size == "large":
            num_particles = 50
            speed_range = (80, 250)
            size_range = (4, 10)
            lifetime_range = (0.5, 1.0)
        else:  # medium
            num_particles = 35
            speed_range = (60, 200)
            size_range = (3, 7)
            lifetime_range = (0.4, 0.8)
        
        # Create explosion particles
        for _ in range(num_particles):
            angle = random.uniform(0, math.pi * 2)
            speed = random.uniform(*speed_range)
            vx = math.cos(angle) * speed
            vy = math.sin(angle) * speed
            
            # Mix of fire colors
            color = random.choice([
                (255, 200, 0),   # Yellow
                (255, 150, 0),   # Orange
                (255, 100, 0),   # Red-orange
                (255, 50, 0),    # Red
                (200, 200, 200), # White-hot
            ])
            
            lifetime = random.uniform(*lifetime_range)
            particle = Particle(
                x=x,
                y=y,
                vx=vx,
                vy=vy,
                lifetime=lifetime,
                max_lifetime=lifetime,
                size=random.uniform(*size_range),
                color=color,
                fade=True,
                shrink=True
            )
            self.particles.append(particle)
        
        # Add smoke particles
        for _ in range(num_particles // 2):
            angle = random.uniform(0, math.pi * 2)
            speed = random.uniform(20, 80)
            vx = math.cos(angle) * speed
            vy = math.sin(angle) * speed
            
            color = random.choice([
                (80, 80, 80),
                (100, 100, 100),
                (120, 120, 120),
            ])
            
            lifetime = random.uniform(0.5, 1.2)
            particle = Particle(
                x=x,
                y=y,
                vx=vx,
                vy=vy,
                lifetime=lifetime,
                max_lifetime=lifetime,
                size=random.uniform(4, 8),
                color=color,
                fade=True,
                shrink=False,
                gravity=20  # Smoke rises slowly
            )
            self.particles.append(particle)
    
    def create_engine_thrust(self, x: float, y: float, angle: float, power: float = 1.0):
        """Create engine thrust particles"""
        # Create 2-3 particles per frame when engine is active
        num_particles = random.randint(2, 3)
        
        for _ in range(num_particles):
            # Thrust goes opposite to engine direction
            thrust_angle = angle + math.pi
            spread = random.uniform(-0.2, 0.2)
            particle_angle = thrust_angle + spread
            
            speed = random.uniform(80, 150) * power
            vx = math.cos(particle_angle) * speed
            vy = math.sin(particle_angle) * speed
            
            # Blue engine glow
            color = random.choice([
                (100, 150, 255),
                (150, 200, 255),
                (200, 220, 255),
            ])
            
            particle = Particle(
                x=x,
                y=y,
                vx=vx,
                vy=vy,
                lifetime=random.uniform(0.1, 0.3),
                max_lifetime=0.3,
                size=random.uniform(2, 4),
                color=color,
                fade=True,
                shrink=True
            )
            self.particles.append(particle)
    
    def create_damage_sparks(self, x: float, y: float):
        """Create sparks when component takes damage"""
        for _ in range(10):
            angle = random.uniform(0, math.pi * 2)
            speed = random.uniform(80, 150)
            vx = math.cos(angle) * speed
            vy = math.sin(angle) * speed
            
            color = random.choice([
                (255, 255, 0),   # Yellow sparks
                (255, 200, 0),   # Orange sparks
                (255, 255, 255), # White sparks
            ])
            
            particle = Particle(
                x=x,
                y=y,
                vx=vx,
                vy=vy,
                lifetime=random.uniform(0.2, 0.4),
                max_lifetime=0.4,
                size=random.uniform(1, 3),
                color=color,
                fade=True,
                shrink=True,
                gravity=200  # Sparks fall
            )
            self.particles.append(particle)
    
    def clear(self):
        """Remove all particles"""
        self.particles.clear()

File: Particles.cs
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using System;
using System.Collections.Generic;

namespace Subspace;

/// <summary>
/// A single particle
/// </summary>
public class Particle
{
    public float X { get; set; }
    public float Y { get; set; }
    public float VX { get; set; }
    public float VY { get; set; }
    public float Lifetime { get; set; }
    public float MaxLifetime { get; set; }
    public float Size { get; set; }
    public Color Color { get; set; }
    public bool Fade { get; set; }
    public bool Shrink { get; set; }
    public float Gravity { get; set; }

    public Particle(float x, float y, float vx, float vy, float lifetime, float size, Color color, 
                   bool fade = true, bool shrink = false, float gravity = 0f)
    {
        X = x;
        Y = y;
        VX = vx;
        VY = vy;
        Lifetime = lifetime;
        MaxLifetime = lifetime;
        Size = size;
        Color = color;
        Fade = fade;
        Shrink = shrink;
        Gravity = gravity;
    }
}

/// <summary>
/// Manages all particles in the game
/// </summary>
public class ParticleSystem
{
    private const float PARTICLE_DRAG = 0.98f;
    private List<Particle> particles = new List<Particle>();
    private Random random = new Random();

    public void Update(float dt)
    {
        // Iterate backwards to safely remove particles
        for (int i = particles.Count - 1; i >= 0; i--)
        {
            var particle = particles[i];
            particle.Lifetime -= dt;

            if (particle.Lifetime <= 0)
            {
                particles.RemoveAt(i);
                continue;
            }

            // Update position
            particle.X += particle.VX * dt;
            particle.Y += particle.VY * dt;

            // Apply gravity
            if (particle.Gravity != 0)
                particle.VY += particle.Gravity * dt;

            // Apply drag
            particle.VX *= PARTICLE_DRAG;
            particle.VY *= PARTICLE_DRAG;
        }
    }

    public void Render(SpriteBatch spriteBatch, Texture2D pixelTexture, float cameraX, float cameraY, int screenWidth, int screenHeight)
    {
        foreach (var particle in particles)
        {
            int screenX = (int)(particle.X - cameraX);
            int screenY = (int)(particle.Y - cameraY);

            // Skip if off screen
            if (screenX < -10 || screenX > screenWidth + 10 || screenY < -10 || screenY > screenHeight + 10)
                continue;

            // Calculate alpha based on lifetime
            float alpha = 1.0f;
            if (particle.Fade)
                alpha = particle.Lifetime / particle.MaxLifetime;

            // Calculate size
            float size = particle.Size;
            if (particle.Shrink)
                size = particle.Size * (particle.Lifetime / particle.MaxLifetime);

            // Render particle
            Color color = particle.Color * alpha;
            int radius = (int)size;

            for (int y = -radius; y <= radius; y++)
            {
                for (int x = -radius; x <= radius; x++)
                {
                    if (x * x + y * y <= radius * radius)
                    {
                        spriteBatch.Draw(pixelTexture, new Rectangle(screenX + x, screenY + y, 1, 1), color);
                    }
                }
            }
        }
    }

    public void CreateWeaponFireEffect(float x, float y, float angle, string weaponType)
    {
        if (weaponType == "laser")
        {
            // Laser muzzle flash - bright and quick
            for (int i = 0; i < 8; i++)
            {
                float speed = 30f + (float)random.NextDouble() * 30f;
                float spread = -0.3f + (float)random.NextDouble() * 0.6f;
                float particleAngle = angle + spread;
                float vx = (float)Math.Cos(particleAngle) * speed;
                float vy = (float)Math.Sin(particleAngle) * speed;

                Color[] colors = { new Color(255, 100, 100), new Color(255, 150, 150), new Color(255, 200, 200) };
                Color color = colors[random.Next(colors.Length)];

                particles.Add(new Particle(x, y, vx, vy, 0.2f, 2f + (float)random.NextDouble() * 2f, color, true, true));
            }
        }
        else  // cannon
        {
            // Cannon smoke and fire
            for (int i = 0; i < 15; i++)
            {
                float speed = 20f + (float)random.NextDouble() * 30f;
                float spread = -0.5f + (float)random.NextDouble() * 1.0f;
                float particleAngle = angle + spread;
                float vx = (float)Math.Cos(particleAngle) * speed;
                float vy = (float)Math.Sin(particleAngle) * speed;

                Color color;
                if (random.NextDouble() < 0.6)
                {
                    Color[] fireColors = { new Color(255, 200, 0), new Color(255, 150, 0), new Color(255, 100, 0) };
                    color = fireColors[random.Next(fireColors.Length)];
                }
                else
                {
                    Color[] smokeColors = { new Color(100, 100, 100), new Color(150, 150, 150) };
                    color = smokeColors[random.Next(smokeColors.Length)];
                }

                particles.Add(new Particle(x, y, vx, vy, 0.3f + (float)random.NextDouble() * 0.2f, 
                    3f + (float)random.NextDouble() * 3f, color, true, false));
            }
        }
    }

    public void CreateExplosion(float x, float y, string size = "medium")
    {
        int numParticles;
        float minSpeed, maxSpeed, minSize, maxSize, minLifetime, maxLifetime;

        switch (size)
        {
            case "small":
                numParticles = 20;
                minSpeed = 50f; maxSpeed = 150f;
                minSize = 2f; maxSize = 5f;
                minLifetime = 0.3f; maxLifetime = 0.6f;
                break;
            case "large":
                numParticles = 50;
                minSpeed = 80f; maxSpeed = 250f;
                minSize = 4f; maxSize = 10f;
                minLifetime = 0.5f; maxLifetime = 1.0f;
                break;
            default:  // medium
                numParticles = 35;
                minSpeed = 60f; maxSpeed = 200f;
                minSize = 3f; maxSize = 7f;
                minLifetime = 0.4f; maxLifetime = 0.8f;
                break;
        }

        // Create explosion particles
        Color[] explosionColors = { 
            new Color(255, 200, 0), new Color(255, 150, 0), new Color(255, 100, 0), 
            new Color(255, 50, 0), new Color(200, 200, 200) 
        };

        for (int i = 0; i < numParticles; i++)
        {
            float angle = (float)(random.NextDouble() * Math.PI * 2);
            float speed = minSpeed + (float)random.NextDouble() * (maxSpeed - minSpeed);
            float vx = (float)Math.Cos(angle) * speed;
            float vy = (float)Math.Sin(angle) * speed;
            Color color = explosionColors[random.Next(explosionColors.Length)];
            float lifetime = minLifetime + (float)random.NextDouble() * (maxLifetime - minLifetime);
            float particleSize = minSize + (float)random.NextDouble() * (maxSize - minSize);

            particles.Add(new Particle(x, y, vx, vy, lifetime, particleSize, color, true, true));
        }

        // Add smoke particles
        Color[] smokeColors = { new Color(80, 80, 80), new Color(100, 100, 100), new Color(120, 120, 120) };
        for (int i = 0; i < numParticles / 2; i++)
        {
            float angle = (float)(random.NextDouble() * Math.PI * 2);
            float speed = 20f + (float)random.NextDouble() * 60f;
            float vx = (float)Math.Cos(angle) * speed;
            float vy = (float)Math.Sin(angle) * speed;
            Color color = smokeColors[random.Next(smokeColors.Length)];
            float lifetime = 0.5f + (float)random.NextDouble() * 0.7f;

            particles.Add(new Particle(x, y, vx, vy, lifetime, 4f + (float)random.NextDouble() * 4f, 
                color, true, false, 20f));
        }
    }

    public void CreateEngineTrust(float x, float y, float angle, float power = 1.0f)
    {
        // Create 2-3 particles per frame when engine is active
        int numParticles = 2 + random.Next(2);

        for (int i = 0; i < numParticles; i++)
        {
            // Thrust goes opposite to engine direction
            float thrustAngle = angle + (float)Math.PI;
            float spread = -0.2f + (float)random.NextDouble() * 0.4f;
            float particleAngle = thrustAngle + spread;

            float speed = (80f + (float)random.NextDouble() * 70f) * power;
            float vx = (float)Math.Cos(particleAngle) * speed;
            float vy = (float)Math.Sin(particleAngle) * speed;

            // Blue engine glow
            Color[] colors = { new Color(100, 150, 255), new Color(150, 200, 255), new Color(200, 220, 255) };
            Color color = colors[random.Next(colors.Length)];

            particles.Add(new Particle(x, y, vx, vy, 0.1f + (float)random.NextDouble() * 0.2f, 
                2f + (float)random.NextDouble() * 2f, color, true, true));
        }
    }

    public void CreateDamageSparks(float x, float y)
    {
        for (int i = 0; i < 10; i++)
        {
            float angle = (float)(random.NextDouble() * Math.PI * 2);
            float speed = 80f + (float)random.NextDouble() * 70f;
            float vx = (float)Math.Cos(angle) * speed;
            float vy = (float)Math.Sin(angle) * speed;

            Color[] colors = { Color.Yellow, new Color(255, 200, 0), Color.White };
            Color color = colors[random.Next(colors.Length)];

            particles.Add(new Particle(x, y, vx, vy, 0.2f + (float)random.NextDouble() * 0.2f, 
                1f + (float)random.NextDouble() * 2f, color, true, true, 200f));
        }
    }

    public void Clear()
    {
        particles.Clear();
    }
}
